Add columns as parameter to get_dataframe

BloodyHell opened the pull request:
Allow the user to specify which columns to retrieve

🎉 Hello there 🎉! Thanks for taking the time to contribute to `Tiingo-Python`!

## Summary of Changes

- What is the motivation / context for this change?
-- I want to retrieve historical information including volume, but the official API specifies that to retrieve the volume column, it needs to be specified explicitly

- Is this a bugfix or a feature?
-- Feature

- What does this change / fix? (Link to Github Issue page if applicable)
-- It adds the option for users to include the 'columns' query parameter

## Checklist

- [x] Code follows the style guidelines of this project
- [x] Added tests for new functionality
- [x] Update `HISTORY.rst` with an entry summarizing your change
- [x] Tag a project maintainer

BloodyHell commented:
@hydrosquall

hydrosquall commented:
Hi @BloodyHell , thanks for making this contribution. I see from the [API docs](https://www.tiingo.com/documentation/end-of-day) that this is indeed a new request parameter. I have a few questions for you.

1. Do you think we should have this param added to the `get_ticker_price` method as well?
2. In order to test this change, we should make sure we have a new unit test, and confirm with a live example that the response in fact changes when we pass at least 2 columns. Do you know what columns are available besides `volume` that we can use to test this? We can list some of the choices in the docstring.

BloodyHell commented:
Fair points @hydrosquall.

1. I've added the variable to `get_ticker_price` as well. Some of that logic could possibly be reused, but that's a change for another moment 
2. I've added two tests, one which requests all columns (all defaults ones including `volume`) and one only requesting `volume`. By default you always get `date`. Also tried to add nice docstring.

hydrosquall commented on HISTORY.rst:
```diff
@@ -5,6 +5,7 @@ History
 --------------------------------
 
 * Dev: New config for readthedocs
+* Feature: Add 'columns' parameter to 'get_dataframe' func (#1057)
```
This can include a mention for `get_ticker_price` too.

BloodyHell replied:
Done

hydrosquall commented on tests/test_tiingo.py:
```diff
@@ -98,6 +98,34 @@ def test_ticker_price_with_csv(self):
         assert len(rows) > 2  # more than 1 day of data
 
+    @vcr.use_cassette('tests/fixtures/ticker_price_with_volume_column.yaml')
+    def test_ticker_price_with_volume_column(self):
```
Good descriptive test names 👍

BloodyHell replied:
I saw I sloppy pasted the description of the tests itself, fixed it in the latest commit :)

hydrosquall commented on tests/test_tiingo_pandas.py:
```diff
@@ -114,6 +114,24 @@ def test_intraday_ticker_price(self):
                                             frequency="30Min")
         self.assertGreater(len(prices), 1)
 
+    @vcr.use_cassette('tests/fixtures/ticker_price_with_volume_column.yaml')
```
Confirming my understanding these two tests makes sense because "date" is always provided as a column, so the # of columns is always the length of `len(columns.split(",")) + 1` .

BloodyHell replied:
You are correct, you always receive the `date` column

BloodyHell replied:
I've updated the _pandas test with your suggestion

hydrosquall replied:
Updated docstrings and variable names make sense, thank you

hydrosquall reviewed (approved):
The change looks good, thank you for making this update + providing tests for both cases!

I've left some optional feedback if you'd like to make any last changes, but otherwise I can take care of that when we cut a new release. 

If you'd like we can add a "volume" usage example to the README.rst at the repository root so people can make use of your parameter, but otherwise people will find out about this param from checking the docstring in their IDE.

BloodyHell commented:
@hydrosquall I've processed your comments and also expanded the README a little bit. Feel free to make changed to it or propose changes.

hydrosquall commented on README.rst:
```diff
@@ -154,11 +157,21 @@ To receive results in ``pandas`` format, use the ``get_dataframe()`` method:
 methods.  Weekly frequencies resample to the end of day on Friday, monthly frequencies resample to the last day of the month, and annually frequencies resample to the end of
 day on 12-31 of each year.  The intraday frequencies are specified using an integer followed by "Min" or "Hour", for example "30Min" or "1Hour".
 
+It's also possible to specify which columns you're interested in, for example:  "open", "close", "low", "high" and "volume" (see https://www.tiingo.com/documentation/iex for future columns).
```
I think the documentation endpoint for this link is https://www.tiingo.com/documentation/end-of-day rather than iex, but I can fix this after I merge your PR.

hydrosquall reviewed (approved):
Thanks for the testing + documentation improvements 👍

hydrosquall reviewed (changes requested):
It looks like something might be off with the new unit tests - can you look into what might have happened?

https://github.com/hydrosquall/tiingo-python/actions/runs/11645778111/job/32429181309?pr=1057

BloodyHell commented:
> It looks like something might be off with the new unit tests - can you look into what might have happened?
> 
> https://github.com/hydrosquall/tiingo-python/actions/runs/11645778111/job/32429181309?pr=1057

Got it. Requesting it as pandas always converts date to the index of the dataframe, thus it doesn't show up in the columns. Expanded the test_tiingo.py case for multiple columns in json format.

hydrosquall commented on tests/test_tiingo_pandas.py:
```diff
@@ -114,6 +114,26 @@ def test_intraday_ticker_price(self):
+                                               columns=requested_column,
+                                               fmt='json')
+        assert len(prices) == 1
+        assert len(prices.columns) == len(requested_column) + 1
```
I don't think we meant to measure the length of `requested_column` for this test. That's the number of characters in the word `volume`,  not the # of columns in the dataframe.

BloodyHell replied:
You are right of course. Python is not my first language, but this is some beginner stuff I missed hahaha

hydrosquall reviewed (changes requested):
We're getting closer, I think there's one more test to fix.

https://github.com/hydrosquall/tiingo-python/actions/runs/11653883333/job/32777625943?pr=1057

hydrosquall reviewed (approved):
We are good to go! Thank you very much for your contribution, as well as for improving documentation and testing along the way.
